RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducer.cc

```diff
@@ -0,0 +1,83 @@
+
+#define DEBUG false
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  class PFClusterSoAProducer : public stream::EDProducer<> {
+  public:
+    PFClusterSoAProducer(edm::ParameterSet const& config)
+        : pfClusParamsToken(esConsumes(config.getParameter<edm::ESInputTag>("pfClusterParams"))),
+          inputPFRecHitSoA_Token_{consumes(config.getParameter<edm::InputTag>("PFRecHitsLabelIn"))},
+          outputPFClusterSoA_Token_{produces()},
+          outputPFRHFractionSoA_Token_{produces()},
+          synchronise_(config.getParameter<bool>("synchronise")),
+          produceSoA_{config.getParameter<bool>("produceSoA")} {}
+
+    void produce(device::Event& event, device::EventSetup const& setup) override {
+      const reco::PFClusterParamsDeviceCollection& params = setup.getData(pfClusParamsToken);
+      const reco::PFRecHitHostCollection& pfRecHits = event.get(inputPFRecHitSoA_Token_);
+      const int nRH = pfRecHits->size();
+
+      reco::PFClusteringVarsDeviceCollection pfClusteringVars{nRH + 1, event.queue()};
+      reco::PFClusteringEdgeVarsDeviceCollection pfClusteringEdgeVars{(nRH * 8) + 1, event.queue()};
+      reco::PFClusterDeviceCollection pfClusters{nRH, event.queue()};
+      reco::PFRecHitFractionDeviceCollection pfrhFractions{nRH * 120, event.queue()};
+
+      PFClusterProducerKernel kernel(event.queue(), pfRecHits);
+      kernel.execute(event.device(),
+                     event.queue(),
+                     params,
+                     pfClusteringVars,
+                     pfClusteringEdgeVars,
+                     pfRecHits,
+                     pfClusters,
+                     pfrhFractions);
+
+      if (synchronise_)
+        alpaka::wait(event.queue());
+
+      if (produceSoA_) {
+        event.emplace(outputPFClusterSoA_Token_, std::move(pfClusters));
+        event.emplace(outputPFRHFractionSoA_Token_, std::move(pfrhFractions));
+      }
+    }
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+      desc.add<edm::InputTag>("PFRecHitsLabelIn");
+      desc.add<std::string>("PFClustersGPUOut", "");
+      desc.add<bool>("produceSoA", true);
+      desc.add<bool>("produceLegacy", true);
+      desc.add<edm::ESInputTag>("pfClusterParams");
+      desc.add<bool>("synchronise");
+      descriptions.addWithDefaultLabel(desc);
+    }
+
+  private:
+    const device::ESGetToken<reco::PFClusterParamsDeviceCollection, JobConfigurationGPURecord> pfClusParamsToken;
+    const edm::EDGetTokenT<reco::PFRecHitHostCollection> inputPFRecHitSoA_Token_;
+    const device::EDPutToken<reco::PFClusterDeviceCollection> outputPFClusterSoA_Token_;
+    const device::EDPutToken<reco::PFRecHitFractionDeviceCollection> outputPFRHFractionSoA_Token_;
+    const bool synchronise_;
+    const bool produceSoA_;
+    int nRH = 0;
+    std::optional<PFClusterProducerKernel> kernel = {};
```

Comment:
`kernel` is a local variable inside `produce()`, I think you can remove this one